arena.go

```diff
@@ -47,55 +45,90 @@ type Arena interface {
 	Release()
 }
 
+// singleSegmentPool is a pool of *SingleSegmentArena.
+var singleSegmentPool = sync.Pool{
+	New: func() any {
+		return &SingleSegmentArena{}
+	},
+}
+
 // SingleSegmentArena is an Arena implementation that stores message data
 // in a continguous slice.  Allocation is performed by first allocating a
 // new slice and copying existing data. SingleSegment arena does not fail
 // unless the caller attempts to access another segment.
-type SingleSegmentArena []byte
+type SingleSegmentArena struct {
+	seg Segment
+
+	// bp is the bufferpool assotiated with this arena if it was initialized
+	// for writing.
+	bp *bufferpool.Pool
+
+	// fromPool determines if this should return to the pool when released.
+	fromPool bool
+}
+
+func zeroSlice(b []byte) {
+	for i := range b {
+		b[i] = 0
+	}
+}
 
 // SingleSegment constructs a SingleSegmentArena from b.  b MAY be nil.
 // Callers MAY use b to populate the segment for reading, or to reserve
 // memory of a specific size.
-func SingleSegment(b []byte) *SingleSegmentArena {
-	return (*SingleSegmentArena)(&b)
+func SingleSegment(b []byte) Arena {
+	if b == nil {
+		ssa := singleSegmentPool.Get().(*SingleSegmentArena)
+		ssa.fromPool = true
+		ssa.bp = &bufferpool.Default
```

Comment:
I'm not sure I understand the interaction between `singleSegmentPool` and `bufferpool.Default`.  Why are we fetching from `singleSegmentPool` and returning to `bufferpool.Default`?

Reply:
`(single|multi)SegmentPool` holds values of their corresponding arena, while `bufferpool.Default` caches `[]byte`.

In other words, the same `[]byte` could be used in either a single or multi segment arena, while the reference to the arena itself may be reused (with different backing slices).

Reply:
Got it -- I'm tracking now. Thanks!